filament/backend/include/backend/Program.h

```diff
@@ -117,6 +120,8 @@ class Program {
     Program& specializationConstants(
             utils::FixedCapacityVector<SpecializationConstant> specConstants) noexcept;
 
+    Program& pushConstants(PushConstantStruct const& pushConstants) noexcept;
```

Comment:
Here we need to describe the push constant buffer, so we shouldn't pass any "value".

I think this should be a vector of names (this will associate the name to an index basically, and it'll only be needed for the gl driver in theory, and maybe the vk driver to allocate the buffer size?).

Reply:
In the definition of the standalone uniform struct for GL:
```
struct MyConstants {
    float t;
};
uniform MyConstants k;
```
Is it ok for `k` remain be the same across all shaders?

Reply:
Yes, I think so.

In gl, you can look up the names of each field of the structure, it will return a location you can use with glUniform to set it.

So what we need, is for Program to be given the name of each field of the structure. I think I'm this example it would be "k.foo", along with an index to be used when setting the value with the driver API (bit I think I'm this case the index should be implicit, based on the order on the vector, which should match the order in the struct{}).

The implementation of program in each backend will then maintain a mapping of its choosing to implement the setPushConstant.

I think on vk/mtl it'll be just an offset into the buffer (like index * 4). In gl it will be mapping from index to location + a call to glUniform.

I hope this makes sense.

Reply:
sure, but one last question.  For `RenderPass` or `PostProcessManager` when they call `driver.setPushConstant(stage, index, value)`, who do they ask for the index (I assume they would only know the name of the variable within the struct), is it `FMaterial`?

Reply:
What I have envisioned for now is that the `index` would be hardcoded inside `EngineEnums.h`. We'd probably have something like:

```
enum class PushConstantId {
    MORPHING_BUFFER_OFFSET = 0,   // uint32
    SRGB_CONVERSION = 1,                   // bool
};
```
In `libfilamat` we would output the push constant block, something like this:

```
struct {
 int morphingBufferOffset;
 bool srgbConversion;
} pushConstants;
```

`MaterialBuilder` would gain a way to define the push constants e.g.:
```
MaterialBuilder::pushConstant(name, type, index);
```
This info will be used to generate the `pushConstants` struct above.

And I think for now the calls to `MaterialBuilder` would just be hardcoded, e.g.:
```
builder.pushConstant("morphingBufferOffset", Type::INT, PushConstantId::MORPHING_BUFFER_OFFSET);
builder.pushConstant("srgbConversion", Type::BOOL, PushConstantId::SRGB_CONVERSION);
```


I *think* the `stage` parameter should only appear when setting the push constant in the driver API, either that, or we assume that they're always set in both -- I don't think we should have a stage per push constant.  In GL they will always be set in both (so they cannot share an id).

Reply:
done